Implementar verificação de disponibilidade de artigos no site QA e public (opac_5)

samuelveigarangel opened the pull request:
#### O que esse PR faz?
Este PR implementa uma funcionalidade automatizada que verifica a disponibilidade dos artigos tanto na versão nova quanto na versão antiga do SciELO. Após a verificação, os resultados são armazenados em modelos específicos e disponibilizados na interface administrativa para consulta.


#### Onde a revisão poderia começar?
Pelos commit https://github.com/scieloorg/scms-upload/commit/b8901abaf9461316fce916a78e679acf24ae5e8a

#### Como este poderia ser testado manualmente?
1. make django_bash
```console
python manage.py runscript load_check_article --script-args {username} 
```
Obs: É importante registrar coleção e periódico e realizar relacionamento com um artigo.

#### Algum cenário de contexto que queira dar?
N/A

### Screenshots
![image](https://github.com/scieloorg/scms-upload/assets/82840278/432788a5-c161-400b-b215-d7006328c3bb)


#### Quais são tickets relevantes?
#454 

### Referências
N/A

samuelveigarangel committed:
Coloca _get_user em um lugar comum aos aplicativos

samuelveigarangel committed:
realiza mudanca do _get_user

samuelveigarangel committed:
realiza mudanca do _get_user

samuelveigarangel committed:
realiza mudanca do _get_user

samuelveigarangel committed:
realiza mudanca do _get_user

samuelveigarangel committed:
realiza mudanca do _get_user e remove _get_user em task_publish_model_inline inutilizavel

samuelveigarangel committed:
realiza mudanca do _get_user

samuelveigarangel committed:
Cria os modelos e seus metodos de classe (CheckArticleAvailability, ScieloSiteStatus)

samuelveigarangel committed:
Insere o paramentro date nos metodos das classes CheckArticleAvailability e ScieloSiteStatus

samuelveigarangel committed:
Cria tasks que realizam a verificacao da disponibilidade do artigo no site

samuelveigarangel committed:
Cria ScieloSiteStatusAdmin

samuelveigarangel committed:
script para executar initiate_article_availability_check

samuelveigarangel committed:
migracao

samuelveigarangel committed:
Corrige for em initiate_article_availability_check

robertatakenaka commented on article/tasks.py:
```diff
@@ -0,0 +1,89 @@
+def process_article_availability(self, user_id, username, pid_v3, journal_acron, lang):
+    urls = [
+        f"https://www.scielo.br/scielo.php?script=sci_arttext&pid={pid_v3}&lng={lang}&nrm=iso", 
+        f"https://www.scielo.br/j/{journal_acron}/a/{pid_v3}/?lang={lang}"
```
@samuelveigarangel a parte www.scielo.br deve ser obtida da class Collection, no entanto, acho que ela está bem diferente da que está no Core. Será necessário atualizá-la

robertatakenaka replied:
@samuelveigarangel adicionar também os formatos pdf

robertatakenaka replied:
@samuelveigarangel acho que seria interessante colocar como melhoria o tempo de resposta

samuelveigarangel replied:
@robertatakenaka Verdade. Acho que posso conseguir esse dado em: https://github.com/scieloorg/scms-upload/blob/91234dd4f0e521e74270df4a9b21c935071cdd0e/collection/models.py#L65

robertatakenaka commented on article/tasks.py:
```diff
@@ -0,0 +1,89 @@
+                response = fetch_data(url, timeout=2, verify=True)
+                CheckArticleAvailability.create_or_update(
+                    article=article,
+                    status=True,
```
@samuelveigarangel trocar True / False por HTTP ERROR CODE, ou pode ser um outro atributo

robertatakenaka reviewed (commented):
@samuelveigarangel Muito bom!

samuelveigarangel committed:
Cria Choices VERIFY_ARTICLE_TYPE E VERIFY_HTTP_ERROR_CODE

samuelveigarangel committed:
- Adiciona campo type e available em ScieloSiteStatus
- Modifica o paramentro Status para exemplicar melhor o error
-  Adiciona paramentro available

samuelveigarangel committed:
- Adiciona bloco try-except em initiate_article_availability_check e process_article_availability
- Melhora o registro das requisicoes em process_article_availability

samuelveigarangel committed:
Adiciona novos displays e filter em ScieloSiteStatus

samuelveigarangel committed:
migracao

robertatakenaka commented on article/tasks.py:
```diff
@@ -0,0 +1,113 @@
+                        pid_v2=article.sps_pkg.articleproc_set.first().pid,
+                        journal_acron=article.journal.journal_acron,
+                        lang=article_per_lang,
+                        domain=article.journal.journalproc_set.first().collection.websiteconfiguration_set.get(enabled=True).url,
```
@samuelveigarangel o teste tem que ser feito para todas as coleções registradas

robertatakenaka commented on article/models.py:
```diff
@@ -314,3 +319,198 @@ def __str__(self) -> str:
+        Modelo para armazenar o status de disponibilidade nos sites,
+        tanto na nova versao, quanto na antiga, do scielo.br.
+    """
+    article = models.ForeignKey(
```
@samuelveigarangel adicione a collection de modo que checará a disponibilidade em todas as coleções. Também deveria ser considerado a disponibilidade no site QA, inclusive saber se está disponível em ambos ou em somente um

robertatakenaka commented on article/choices.py:
```diff
@@ -124,3 +125,13 @@
     (AS_PUBLISHED, _("Published")),
 )
+
+VERIFY_ARTICLE_TYPE = [
```
@samuelveigarangel nome não adequado, melhor FORMAT no lugar de ARTICLE_TYPE

robertatakenaka commented on article/models.py:
```diff
@@ -29,6 +30,10 @@
 User = get_user_model()
 
 
+def verify_type_of_url(type):
```
@samuelveigarangel qual é a necessidade desta função?

samuelveigarangel replied:
Não há necessidade de verificar a URL  do pdf do artigo?

robertatakenaka commented on article/models.py:
```diff
@@ -314,3 +319,198 @@ def __str__(self) -> str:
     base_form_class = RequestArticleChangeForm
+
+
+class CheckArticleAvailability(CommonControlField):
```
@samuelveigarangel Use `ArticleAvailability` no lugar de `CheckArticleAvailability`

robertatakenaka replied:
@samuelveigarangel  mova este modelo para dentro de publication, pois dentro de publication podemos importar quaisquer outros modelos como JournalProc, IssueProc e outros minimizando problemas de dependências circulares

robertatakenaka commented on article/models.py:
```diff
@@ -314,3 +319,198 @@ def __str__(self) -> str:
+        unique=True,
+    )
+    site_status = models.ManyToManyField(
+        "ScieloSiteStatus"
```
@samuelveigarangel mude para url_status e "URLStatus

robertatakenaka commented on article/models.py:
```diff
@@ -314,3 +319,198 @@ def __str__(self) -> str:
+            )
+
+class ScieloSiteStatus(CommonControlField):
+    check_date = models.DateTimeField(null=True, blank=True)
```
@samuelveigarangel precisa de check_date se há updated?

robertatakenaka commented on article/models.py:
```diff
@@ -314,3 +319,198 @@ def __str__(self) -> str:
+
+class ScieloSiteStatus(CommonControlField):
+    check_date = models.DateTimeField(null=True, blank=True)
+    url_site_scielo = models.SlugField(max_length=500, unique=True)
```
@samuelveigarangel por que SlugField no lugar de URLField?

robertatakenaka commented on article/models.py:
```diff
@@ -314,3 +319,198 @@ def __str__(self) -> str:
+class ScieloSiteStatus(CommonControlField):
+    check_date = models.DateTimeField(null=True, blank=True)
+    url_site_scielo = models.SlugField(max_length=500, unique=True)
+    status = models.CharField(
```
@samuelveigarangel use HTTPCode (SmallPositiveInteger)

robertatakenaka commented on article/models.py:
```diff
@@ -314,3 +319,198 @@ def __str__(self) -> str:
+        null=True, 
+        blank=True
+    )
+    type = models.CharField(
```
@samuelveigarangel precisa desta distinção?

robertatakenaka commented on article/models.py:
```diff
@@ -314,3 +319,198 @@ def __str__(self) -> str:
+        null=True, 
+        blank=True,
+    )
+    available = models.BooleanField(default=False)
```
@samuelveigarangel qual é a diferença entre status e available?

samuelveigarangel replied:
Status pode dar uma mensagem melhor para o usuário. Distingué entre uma URL errada ou instabilidade na URL. Talvez o nome do campo possa ser melhor.

robertatakenaka commented on article/models.py:
```diff
@@ -314,3 +319,198 @@ def __str__(self) -> str:
+        return self
+
+    class Meta:
+        verbose_name = "Scielo Site Status"
+        verbose_name_plural = "Scielo Site Status"
```
URL status

robertatakenaka commented on article/tasks.py:
```diff
@@ -0,0 +1,113 @@
+import re
```
@samuelveigarangel mover este conteúdo para publication/tasks.py

robertatakenaka commented on article/wagtail_hooks.py:
```diff
@@ -14,7 +14,7 @@
 
 from .button_helper import ArticleButtonHelper, RequestArticleChangeButtonHelper
-from .models import Article, RelatedItem, RequestArticleChange, choices
+from .models import Article, RelatedItem, RequestArticleChange, choices, ScieloSiteStatus
```
@samuelveigarangel mover este conteúdo para publication/wagtail_hooks.py

robertatakenaka reviewed (changes requested):
@samuelveigarangel separar em 2 PRs: refatoração do get_user e solução da verificação da disponibilidade do site

samuelveigarangel committed:
Merge branch 'main' into issue-454

samuelveigarangel commented:
@robertatakenaka, vou fechar esta PR e abrir uma nova com o repositório atualizado, pois esta branch está bastante desatualizada.
